src/cdk/v2/destinations/intercom/utils.js

```diff
@@ -0,0 +1,252 @@
+  isHttpStatusSuccess,
+  getFieldValueFromMessage,
+  removeUndefinedAndNullValues,
+} = require('../../../../v0/util');
+const { JSON_MIME_TYPE } = require('../../../../v0/util/constant');
+const {
+  BASE_ENDPOINT,
+  MetadataTypes,
+  BASE_EU_ENDPOINT,
+  BASE_AU_ENDPOINT,
+  ReservedUserAttributes,
+  SEARCH_CONTACT_ENDPOINT,
+  ReservedCompanyAttributes,
+  CREATE_OR_UPDATE_COMPANY_ENDPOINT,
+} = require('./config');
+
+/**
+ * Returns destination request headers
+ * @param {*} destination
+ * @returns
+ */
+const getHeaders = (destination) => ({
+  'Content-Type': JSON_MIME_TYPE,
+  Authorization: `Bearer ${destination.Config.apiKey}`,
+  Accept: JSON_MIME_TYPE,
+  'Intercom-Version': '2.10',
+});
+
+/**
+ * Returns destination request base endpoint
+ * @param {*} destination
+ * @returns
+ */
+const getBaseEndpoint = (destination) => {
+  const { apiServer } = destination.Config;
+  if (apiServer === 'eu') {
+    return BASE_EU_ENDPOINT;
+  }
+  if (apiServer === 'au') {
+    return BASE_AU_ENDPOINT;
+  }
+  return BASE_ENDPOINT;
+};
+
+/**
+ * Returns contact lookup field
+ * @param {*} message
+ * @returns
+ */
+const getLookUpField = (message) => {
+  let lookupField = 'email';
+  const integrationsObj = getIntegrationsObj(message, 'INTERCOM');
+  if (integrationsObj && isDefinedAndNotNull(integrationsObj.lookup)) {
+    lookupField = integrationsObj.lookup;
+  }
+  return lookupField;
+};
+
+/**
+ * Returns the value of name field
+ * @param {*} message
+ * @returns
+ */
+const getName = (message) => {
+  const name = message?.traits?.name || message?.context?.traits?.name;
+  if (name) return name;
+  const firstName = getFieldValueFromMessage(message, 'firstName');
+  const lastName = getFieldValueFromMessage(message, 'lastName');
+  if (firstName && lastName) {
+    return `${firstName} ${lastName}`;
+  }
+
+  if (firstName || lastName) {
+    return firstName || lastName;
+  }
+  return undefined;
+};
+
+/**
+ * Returns custom attributes for identify and group calls (for contact and company in intercom)
+ * @param {*} payload
+ * @param {*} type
+ * @returns
+ */
+const filterCustomAttributes = (payload, type) => {
+  const ReservedAttributes = type === 'user' ? ReservedUserAttributes : ReservedCompanyAttributes;
+  let { custom_attributes: customAttributes } = payload;
+  if (customAttributes) {
+    ReservedAttributes.forEach((trait) => {
+      if (customAttributes[trait]) delete customAttributes[trait];
+    });
+
+    if (isDefinedAndNotNull(customAttributes) && Object.keys(customAttributes).length > 0) {
+      customAttributes = flattenJson(customAttributes);
+    }
+  }
+  return customAttributes;
+};
+
+/**
+ * Api call to search contact in intercom to returns id of contact
+ * Ref doc : https://developers.intercom.com/docs/references/rest-api/api.intercom.io/Contacts/SearchContacts/
+ * @param {*} message
+ * @param {*} destination
+ * @returns
+ */
+const searchContact = async (message, destination) => {
```

Comment:
Have we checked the limit of the search API?

Reply:
Nesting & Limitations

You can nest these filters in order to get even more granular insights that pinpoint exactly what you need. Example: (1 OR 2) AND (3 OR 4). There are some limitations to the amount of multiple's there can be:

There's a limit of max 2 nested filters
There's a limit of max 15 filters for each AND or OR group

this are their query limits which we are already following

rate limits is same and applicable to all endpoints as per below doc
https://developers.intercom.com/docs/references/rest-api/errors/rate-limiting/